invenio_sipstore/models.py

```diff
@@ -0,0 +1,112 @@
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Invenio-SIPStore database models."""
+
+from __future__ import absolute_import, print_function
+from sqlalchemy_utils.models import Timestamp
+from sqlalchemy_utils.types import JSONType, UUIDType
+from invenio_files_rest.models import FileInstance
+from invenio_accounts.models import User
+from invenio_pidstore.models import PersistentIdentifier
+
+from invenio_db import db
+import uuid
+
+
+class SIP(db.Model, Timestamp):
+    """Submission Information Package model."""
+
+    __tablename__ = 'sipstore_sip'
+
+    id = db.Column(UUIDType, primary_key=True, default=uuid.uuid4)
+    """Id of SIP."""
+
+    sip_format = db.Column(db.String(7), nullable=False)
+    """Format of the SIP content ('json', 'marcxml' etc.)"""
+
+    content = db.Column(db.Text, nullable=False)
+    """Text blob of the SIP content."""
+
+    user_id = db.Column(db.Integer,
+                        db.ForeignKey(User.id),
+                        nullable=True,
+                        default=None)
+    """User responsible for the SIP."""
+
+    agent = db.Column(JSONType, default=lambda: dict(), nullable=False)
+    """Agent information regarding given SIP."""
+
+    #
+    # Relationships
+    #
+    user = db.relationship(User, backref='sips', foreign_keys=[user_id])
+    """Relation to the User responsible for the SIP."""
+
+
+class SIPFile(db.Model, Timestamp):
+    """Extra SIP info regarding files."""
+
+    __tablename__ = 'sipstore_sipfile'
+
+    sip_id = db.Column(UUIDType, db.ForeignKey(SIP.id))
+    """Id of SIP."""
+
+    filename = db.Column(db.String(255), nullable=False)
+    """Filename of submitted file within the SIP record."""
+
+    file_id = db.Column(UUIDType, db.ForeignKey(FileInstance.id),
+                        primary_key=True,
+                        nullable=False)
+    """Id of the FileInstance."""
+
+    #
+    # Relations
+    #
+    sip = db.relationship(SIP, backref='sip_files', foreign_keys=[sip_id])
+    """Relation to the SIP along which given file was submitted."""
+
+    file = db.relationship(FileInstance, backref='sip_files',
+                           foreign_keys=[file_id])
+    """Relation to the SIP along which given file was submitted."""
+
+
+class RecordSIP(db.Model, Timestamp):
+    """An association table for Records and SIPs."""
+
+    __tablename__ = 'sipstore_recordsip'
+
+    sip_id = db.Column(UUIDType, db.ForeignKey(SIP.id),
+                       primary_key=True)
+    """Id of SIP."""
+
+    pid_id = db.Column(db.Integer, db.ForeignKey(PersistentIdentifier.id),
+                       primary_key=True)
+    """Id of the PID pointing to the record."""
```

Comment:
So if I am understanding correctly, a SIP would point to a PID that points to a record. If the given record is merged with another one, since the the PID is moved accordingly, there is no need for the SIP to be updated. Neat!